### ISSUE TITLE
Entity returned from server is left in a Detached state when it references itself

### ISSUE THREAD

**Issue by erestor** [Breeze/breeze-client#87]
https://github.com/Breeze/breeze-client/blob/83264900b6396b1c6530b96c0bb94e3e527252a1/src/mapping-context.ts#L308

When an object referencing itself is returned from the EFCore server, it doesn't correctly use $ref (IMO EFCore bug) and the two objects are in the JSON instead of one, e.g. `{ 'id': 1, 'ownerId': 1, 'owner': { 'id': 1, 'ownerId': 1, 'owner': null, ... }, ... }`. The mapping in Breeze results in entityAspect to have state 'Detached' and an empty entityManager. This is because the reference object is processed first (due to recursion in visitAndMerge) and stored. It is then correctly recognized inside _attachEntityCore and returned, but the return value is disregarded, leaving the original entity being mapped detached.

Suggested fix: replace the line with `targetEntity = em._attachEntityCore(targetEntity, EntityState.Unchanged, mergeStrategy);`

This would be in accord with EntityManager: https://github.com/Breeze/breeze-client/blob/83264900b6396b1c6530b96c0bb94e3e527252a1/src/entity-manager.ts#L797

**Comment by steveschmitt** [comment]
Good catch!  Can you provide an example to reproduce this?

**Comment by erestor** [comment]
I'm afraid I can hardly justify the time to contrive a working server-client example for a one-liner. I'm happy to jump on a call and show you, it shouldn't take more than 10 minutes.

**Comment by steveschmitt** [comment]
No problem; I can try to do it myself on top of our existing test code.  I mostly just want you to clarify the scenario here.
If I have an entity class that references the same class, like this:
```cs
class Employee {
    public int Id { get; set; }
    public int ManagerId  { get; set; }
    public Employee Manager  { get; set; }
}
```
Does the error happen for all Employee-Manager relationships, or only when the Manager is the same entity as the employee and `ManagerId` == `Id` ?

**Comment by erestor** [comment]
Yeah it only happens when the entity references itself, so ManagerId == Id.

(Interestingly it does not happen when using EF 4.8 on the server, only for EF Core. The code in materialization/shaper.cs finds the existing object correctly in ObjectStateManager on line 188 in EF 4.8; I was not able to debug into EF Core in enough detail to uncover the problem, so I didn't report to them.)

**Comment by steveschmitt** [comment]
If it's a bug in EF Core, we are lucky that the fix is so easy in the Breeze client.

**Comment by erestor** [comment]
Further investigation leads to the conclusion that this is when UseQueryTrackingBehavior(QueryTrackingBehavior.NoTracking) is used in C# for the database context, which switches off identity resolution.
Still Breeze should work in this scenario, so the issue stands.

**Comment by steveschmitt** [comment]
I _think_ this is fixed in version 2.1.6.  Can you let me know?

**Comment by erestor** [comment]
Hi, looking at the diff I can't see any code handling this. See #88
